### ISSUE TITLE
Async log interface on leader

### ISSUE THREAD

**Issue by yossigo** [RedisLabs/redisraft#76]
When appending entries to the local log, the leader can handle the operation asynchronously while also sending AppendEntries messages to cut latency down.

**Comment by sjpotter** [comment]
I'm trying to make sense of this, but it seems a bit difficult.

the benefit of this, as I understand it, is that remote and local operations can interleave.  right now, we can interleave the remote operations, but they have to wait till the local operation is done.  if we could interleave them all, we'd presumambly have better performance as remote operations would run while local operation is running.

1) imagine leader does appending local asynch with sending to other nodes

what happens if leader write fails, but other nodes were fine?  currently, this isn't an issue, if leader append fails, we don't send entries to other nodes

So this could be a big showstopper preventing interleaving them, if this breaks a required property of raft.  But let's assume it does not.

2) I think you would have to have a form of locking/blocking, as there's lots of code that expects that only one entry can be in flight at a time on the leader.  i.e. getting current indexes.  I'm wondering if we'd make leader more asynchronous that this would be problematic.

Off the top of my head, I'm wondering if we could really break the asynchronousness of raft_recv_entry().   I'm wondering if a better solution (assuming that the issue I listed in #1 isn't an issue), that we restructure raft_recv_entry() a little.

Currently, we do the local log, then we raft_send_appendentries() to all the nodes.  What about if we switched it around, and moved node communication out of the main redisraft thread, and into another thread.  Then, when we kick off all the raft_send_appendentries() they will start running immediately, as not in our main thread, while the local write can work synchronously and keep raft_recv_entry() synchronous.

as I said, it depends on how this would impact the raft algorithm, but I think anything that makes the leader append asynch to raft_send_appendentries() would have the same breakage.
